src/fqdncache.h

```diff
@@ -11,19 +11,17 @@
 #ifndef SQUID_SRC_FQDNCACHE_H
 #define SQUID_SRC_FQDNCACHE_H
 
-#include "ip/Address.h"
+#include "dns/forward.h"
```

Comment:
Please explicitly include ip/forward.h instead of starting to rely on current side effects of dns/forward.h inclusion (to get a forward declaration of Ip::Address).

```suggestion
#include "dns/forward.h"
#include "ip/forward.h"
```

Reply:
`dns/forward.h` includes `ip/forward.h` , no need to include it twice.

Reply:
> `dns/forward.h` includes `ip/forward.h` 

Yes, my change request is based on this fact.

> no need to include it twice.

To get the current code compiled, we do not need to include it explicitly (as already acknowledged in my change request).

To avoid relying on side effects, we do need to include it explicitly (as already stated in my change request). This reason is enough to include ip/forward.h explicitly despite the current code compiling without that explicit inclusion.

Reply:
I think your definition of "side-effects" in this case is wrongly applied. The `dns/forward.h` provides the declarations for libdns API. The storage type(s) used for DNS responses (eg. `Ip::Address`) are part of that and always will be - the definition(s) will be provided via `dns/forward.h`.

Reply:
> The `dns/forward.h` provides the declarations for libdns API.

dns/forward.h should provide _forward declarations_ for libdns API. Today, it also provides three function declarations, but that (hopefully temporary) violation is not to be relied on when making design decisions (not to mention relying on side effects of that violation!).

> The storage type(s) used for DNS responses (eg. Ip::Address) are part of that and always will be

Ip::Address is not a part of libdns forward declarations today, and even if it becomes a part of them tomorrow, it is not something libdns users should rely on.

The relevant rule of thumb is quite simple: If header B.h uses a name declared in header A.h, then B.h should include either A/forward.h (if sufficient to get B.h compiled) or header A.h itself (otherwise).

Reply:
We are relying on side effects all over the place, and we're not consistent about any rule,
and we also didn't agree on any rule (and we have a lot already).
However, in this case it seems fitting, and it should help future-proof this file.
Finally, I don't feel it's something worth holding up a PR over.

It's been a long while, let's clear this out of our queue